armsrc/legicrf.c

```diff
@@ -40,6 +43,458 @@ static int      legic_reqresp_drift;
+#define RWD_TIME_0           90 /* READER_TIME_PAUSE 20us off + 40us on = 60us */
+#define RWD_FRAME_WAIT      330 /* 220us from TAG frame end to READER frame start */
+#define TAG_FRAME_WAIT      495 /* 330us from READER frame end to TAG frame start */
+#define TAG_BIT_PERIOD      150 /* 100us */
+#define TAG_WRITE_TIMEOUT    60 /* 40 * 100us (write should take at most 3.6ms) */
+
+#define SIM_DIVISOR         586 /* prng_time/DIV count prng needs to be forwared */
+#define SIM_SHIFT           900 /* prng_time+SHIFT shift of delayed start */
+#define RWD_TIME_FUZZ        20 /* rather generous 13us, since the peak detector
+                                /+ hysteresis fuzz quite a bit */
+
+#define LEGIC_READ         0x01 /* Read Command */
+#define LEGIC_WRITE        0x00 /* Write Command */
+
+#define SESSION_IV 0x55         /* An arbitrary chose session IV, all shoud work */
+#define OFFSET_LOG 1024         /* The largest Legic Prime card is 1k */
+#define WRITE_LOWERLIMIT      4 /* UID and MCC are not writable */
+
+#define INPUT_THRESHOLD       8 /* heuristically determined, lower values */
+                                /* lead to detecting false ack during write */
+
+#define FUZZ_EQUAL(value, target, fuzz) ((value) > ((target)-(fuzz)) && (value) < ((target)+(fuzz)))
+
+//-----------------------------------------------------------------------------
+// I/O interface abstraction (FPGA -> ARM)
+//-----------------------------------------------------------------------------
+
+static inline uint8_t rx_byte_from_fpga() {
+  for(;;) {
+    WDT_HIT();
+
+    // wait for byte be become available in rx holding register
+    if(AT91C_BASE_SSC->SSC_SR & (AT91C_SSC_RXRDY)) {
+      return AT91C_BASE_SSC->SSC_RHR;
+    }
+  }
+}
+
+//-----------------------------------------------------------------------------
+// Demodulation (Reader)
+//-----------------------------------------------------------------------------
+
+// Returns am aproximated power measurement
+//
+// The FPGA running on the xcorrelation kernel samples the subcarrier at ~3 MHz.
+// The kernel was initialy designed to receive BSPK/2-PSK. Hance, it reports an
+// I/Q pair every 18.9us (8 bits i and 8 bits q).
+//
+// The subcarrier amplitude can be calculated using Pythagoras sqrt(i^2 + q^2).
+// To reduce CPU time the amplitude is approximated by using linear functions:
+//   am = MAX(ABS(i),ABS(q)) + 1/2*MIN(ABS(i),ABSq))
+//
+// Note: The SSC receiver is never synchronized the calculation my be performed
+// on a i/q pair from two subsequent correlations, but does not matter.
+static inline int32_t sample_power() {
+  int32_t q = (int8_t)rx_byte_from_fpga(); q = ABS(q);
+  int32_t i = (int8_t)rx_byte_from_fpga(); i = ABS(i);
+
+  return MAX(i, q) + (MIN(i, q) >> 1);
+}
+
+// Returns a demedulated bit
+//
+// An aproximated power measurement is available every 18.9us. The bit time
+// is 100us. The code samples 5 times and uses the last (most stable) sample.
```

Comment:
The very first two samples after switching to xcorr mode are garbage because this data has been acqired before the tag started sending. So your fifth sample is in fact in the middle of the tag's answer. Which is good because it allows some slack in timing (and therefore is the most stable as you have stated). It may be worth to think about adding (averaging) over a few samples.

Reply:
![rx dc offset](https://user-images.githubusercontent.com/8958160/44167784-6cde9100-a0cf-11e8-846f-3103df08a6fc.png)

I noticed that it takes up to 30us of the 100us until the Analog signal settels. Hence using the first 3 samples produced more noise than gain.

Reply:
I'm not sure yet how to approach the new Q/I pairs yet I'm considering to start to average 4 of them. However that should happen on the RAW Q/I samples.

Reply:
The impact of the slow settling on the 212kHz correlator should be marginal. It might be worth to add the detected amplitudes to Bigbuf and look at the result with data samples / data plot.

Reply:
Here a sequence of 1 0 0 1 1 1
```
#db# sample    0,    0          
#db# sample   -6, -118          
#db# sample  -16, -124          
#db# sample  -16, -128          
#db# sample  -16, -128          
#db# sample 88888888, 88888888          
#db# sample -107,  -16          
#db# sample   -9,  -12          
#db# sample    0,   -3          
#db# sample   -1,    0          
#db# sample    0,   -1          
#db# sample 88888888, 88888888          
#db# sample    0,    0          
#db# sample    0,    0          
#db# sample    0,    0          
#db# sample   -1,    0          
#db# sample   -1,    0          
#db# sample 88888888, 88888888          
#db# sample    0,    0          
#db# sample   -9, -123          
#db# sample  -17, -126          
#db# sample  -16, -127          
#db# sample  -15, -127          
#db# sample 88888888, 88888888          
#db# sample -127,  -16          
#db# sample -127,  -16          
#db# sample -127,  -16          
#db# sample -127,  -16          
#db# sample -127,  -16          
#db# sample 88888888, 88888888          
#db# sample -127,  -16          
#db# sample -127,  -16          
#db# sample -127,  -16          
#db# sample -127,  -15          
#db# sample -127,  -15          
#db# sample 88888888, 88888888          
```

Reply:
And the same (Proxmark3 Rev1)

```
#db# sample  -44,    0          
#db# sample  -97,  -30          
#db# sample -115, -112          
#db# sample -116, -107          
#db# sample -115, -104          
#db# sample 88888888, 88888888          
#db# sample -116, -103          
#db# sample  -13, -102          
#db# sample  -10,   -1          
#db# sample   -4,   -1          
#db# sample   -1,    0          
#db# sample 88888888, 88888888          
#db# sample    0,    0          
#db# sample    0,    0          
#db# sample    0,    1          
#db# sample   -1,    0          
#db# sample    0,    0          
#db# sample 88888888, 88888888          
#db# sample  -94,    0          
#db# sample -125,  -68          
#db# sample -116, -126          
#db# sample -117, -106          
#db# sample -116, -103          
#db# sample 88888888, 88888888          
#db# sample -115, -102          
#db# sample -115, -102          
#db# sample -116, -103          
#db# sample -115, -102          
#db# sample -115, -102          
#db# sample 88888888, 88888888          
#db# sample -101, -116          
#db# sample -102, -115          
#db# sample -102, -115          
#db# sample -102, -115          
#db# sample -101, -116          
#db# sample 88888888, 88888888  
```

Reply:
The last 2 samples are the best, the rest reduces the clarity of the signal

Reply:
Hmm, I think that several effects are adding up here, mostly caused by the old FPGA code:
1. Old FPGA code averages over 4 subcarrier cycles. If you are unlucky, a Q/I pair covers half 1 and half 0.
2. The whole sequence is off by up to 1.5 Q/I pairs. The old FPGA code needs 4 subcarrier cycles to sample a Q/I pair, and another 2 subcarrier cycles to send the first result (the I value) to the ARM. I.e. when the ARM received a Q/I pair, it is already outdated by up to 28us (!)
3. And finally, the effect of the ramp (which I still think is marginal :smiley:)

Reply:
4. Q and I seem to switch sometimes from bit to bit...

Reply:
I‘ve uploaded samples with the new FPGA.

Reply:
Wow. Looks promising. We clearly see the clipping work. How does it look for long reading distances? Summing up over lets say the last 15 amplitudes should eliminate any noise. Probably the maximum reading distance is now limited by the ability to power up the card...